service/controller/azureconfig/handler/volumebindingmigration/resource.go

```diff
@@ -0,0 +1,130 @@
+package volumebindingmigration
+
+import (
+	"context"
+
+	providerv1alpha1 "github.com/giantswarm/apiextensions/v3/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/k8sclient/v5/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/tenantcluster/v3/pkg/tenantcluster"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/azure-operator/v5/service/controller/key"
+)
+
+const (
+	// Name is the identifier of the resource.
+	Name = "volumebindingmigration"
+)
+
+type Config struct {
+	Logger                   micrologger.Logger
+	TenantRestConfigProvider *tenantcluster.TenantCluster
+}
+
+// Resource ensures that existing StorageClasses use `WaitForFirstConsumer`
+// bind mode.
+type Resource struct {
+	logger                   micrologger.Logger
+	tenantRestConfigProvider *tenantcluster.TenantCluster
+}
+
+func New(config Config) (*Resource, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+	if config.TenantRestConfigProvider == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.TenantRestConfigProvider must not be empty", config)
+	}
+
+	r := &Resource{
+		logger:                   config.Logger,
+		tenantRestConfigProvider: config.TenantRestConfigProvider,
+	}
+
+	return r, nil
+}
+
+// EnsureCreated ensures that existing StorageClasses use
+// `WaitForFirstConsumer` bind mode.
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToCustomResource(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	var tenantClusterK8sClient client.Client
+	{
+		tenantClusterK8sClient, err = r.getTenantClusterClient(ctx, &cr)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	desiredVolumeBindMode := storagev1.VolumeBindingWaitForFirstConsumer
+
+	r.logger.Debugf(ctx, "ensuring storageclasses use desired volumeBindMode %q", desiredVolumeBindMode)
+
+	storageClassList := &storagev1.StorageClassList{}
+	err = tenantClusterK8sClient.List(ctx, storageClassList)
```

Comment:
Maybe we should update only the storage classes we created and not all of them. There might be others created by the customer.

Reply:
Ack. I'll add a filter for those.